### ISSUE TITLE
TRL SGLang Support

### ISSUE THREAD

**Issue by zhaochenyang20** [zhaochenyang20/Awesome-ML-SYS-Tutorial#103]
我们要在 trl 的 GRPO trainer 里面支持 SGLang，目前把这些文档顺着读完就好了。

1. https://zhuanlan.zhihu.com/p/5220718268
2. https://zhuanlan.zhihu.com/p/677607581
3. https://github.com/zhaochenyang20/Awesome-ML-SYS-Tutorial/tree/main/rlhf/OpenRLHF#rlhf-%E7%9A%84%E8%AE%A1%E7%AE%97%E6%B5%81
4. https://zhuanlan.zhihu.com/p/16370000391
5. https://github.com/zhaochenyang20/Awesome-ML-SYS-Tutorial/blob/main/rlhf/OpenRLHF/develop-log.md
6. https://github.com/zhaochenyang20/Awesome-ML-SYS-Tutorial/blob/main/rlhf/GRPO/SGLang_GRPO.md
7. https://github.com/zhaochenyang20/Awesome-ML-SYS-Tutorial/blob/main/rlhf/verl/server-based/veRL-server-based-rollout.md

TODO:
- [x] run trl[vllm] @ryang-max @renxinx 这个都得跑，而且瑞国需要交接好怎么跑 ETA: 4.18
- [ ] implement sglang-serve(server-based sglang for trl, similar to vllm) @ryang-max 这个并不麻烦，非常好写 ETA 4.21
- [ ] implement grpo_config and grpo_train @renxinx @ryang-max ETA 4.23
- [ ] Integration Test @renxinx ETA 4.26
- [ ] Compare performance with vllm @ryang-max @renxinx ETA 4.29

**Comment by zhaochenyang20** [comment]
1. https://docs.sglang.ai/backend/send_request.html
2. https://docs.sglang.ai/backend/openai_api_completions.html
3. https://docs.sglang.ai/backend/native_api.html
4. https://docs.sglang.ai/backend/offline_engine_api.html

@renxinx

**Comment by zhaochenyang20** [comment]
任何问题，首先问 Grok，然后查 google，比如 huggingface blog，一小时自己搞不定就来问人！问人之前，把这个问题写下来阐述清楚，然后发 link 给大家，比如直接写到这个 issue 下面，然后发 link 问人就好了。

**Comment by jhinpan** [comment]
# TRL SGLang GRPO 实现 RoadMap

## 1. 背景与目标

GRPO (Generative Reinforcement Learning with Preference Optimization) 是一种高效的强化学习算法，可用于训练大型语言模型。本项目旨在将 SGLang 作为 inference engine 集成到 Hugging Face 的 TRL (Transformer Reinforcement Learning) 库中，以提高 GRPO 的训练效率和性能。

**主要目标**：
- 实现 SGLang 与 TRL 的无缝集成
- 优化跨 GPU 权重更新机制
- 提高推理速度，降低训练时间

## 2. 系统架构

### 2.1 整体架构

系统采用分布式架构，使用多 GPU 协同工作，以4张卡为例：
- 训练卡（3张）：执行模型训练
- 推理卡（1张）：使用 SGLang 进行高效推理

### 2.2 核心组件

- **训练引擎**：基于 Hugging Face TRL 库
- **推理引擎**：SGLang
- **权重同步机制**：基于 tensor 的跨进程权重更新
- **服务协调**：采用服务器模式管理训练和推理进程

## 3. 主要实现步骤

### 3.1 Initialization and Dedicated GPU Selection

实现专用 GPU 选择机制，确保训练和推理任务分配到合适的 GPU：

Key Points：
- 明确区分训练 GPU 和推理 GPU
- 实现智能资源分配
- 确保不同进程间 GPU 资源不冲突

### 3.2 权重更新机制: update_weights_from_tensor

替换原有的基于磁盘的权重更新方法 (update_weights_from_disk)，实现基于 tensor 的高效权重更新：

优势：
- 避免磁盘 I/O 开销，显著提高权重更新速度
- 适用于单节点多 GPU 环境
- 减少训练中断时间

### 3.3 SGLang 作为 Inference Engine

集成 SGLang 作为高性能推理引擎，加速 rollout 过程：

参考 vllm 服务器实现方式，开发类似的 SGLang 服务器模块：
- 创建独立的 SGLang 服务进程
- 实现权重同步 RPC 接口
- 优化推理参数配置

## 4. 服务器架构重构

参考 Hugging Face 最近的 GRPO vLLM 后端重构，采用基于服务器的架构：

关键改进：
- 使用 SGLang 内部的集体通信机制进行权重同步
- 实现类似 vllm.collective_rpc 的接口
- 优化服务器资源管理

## 5. 测试与验证

### 5.1 性能测试

设计测试方案验证实现效果：
- 对比 update_weights_from_disk 和 update_weights_from_tensor 的性能差异
- 测量 SGLang 推理加速效果
- 评估整体训练时间改进

### 5.2 集成测试

确保各组件协同工作：
- 验证训练与推理进程通信
- 测试权重同步正确性
- 检查资源使用效率

## 6. 参考实现

当前实现基于以下工作：
- 原始仓库：https://github.com/jhinpan/trl-jin/tree/main
- 提交的 PR：https://github.com/huggingface/trl/pull/2981
- Hugging Face 重构 PR：https://github.com/huggingface/trl/pull/3094
- vllm 服务实现：https://github.com/huggingface/trl/blob/main/trl/scripts/vllm_serve.py

@renxinx @ryang-max @zhaochenyang20

**Comment by zhaochenyang20** [comment]
@hebiao064

**Comment by ryang-max** [comment]
上面的文档都看过了，今天开始看vllm server-based的实现，然后先跑一下

**Comment by renxinx** [comment]
> 1. https://docs.sglang.ai/backend/send_request.html
> 2. https://docs.sglang.ai/backend/openai_api_completions.html
> 3. https://docs.sglang.ai/backend/native_api.html
> 4. https://docs.sglang.ai/backend/offline_engine_api.html
> 
> [@renxinx](https://github.com/renxinx)

这些都看过了

**Comment by ryang-max** [comment]
TODO:
1. run trl[vllm] 
2. implement sglang-serve(server-based sglang for trl, similar to vllm) @ryang-max 
3. implement grpo_config and grpo_train @renxinx 
4. Integration Test @ryang-max @renxinx 
5. Compare performance with vllm @ryang-max @renxinx 

cc @zhaochenyang20

**Comment by zhaochenyang20** [comment]
@ryang-max 你直接更新到最上面，不要 follow up

**Comment by zhaochenyang20** [comment]
1. 什么是 transformer 的 forward 计算；
2. 四个角色谁需要被训练，谁需要被 inference，每个人用的 engine 是什么？为什么需要用特定的 engine（为什么不能用 sglang 计算 reference model 的 inference）；
3. 第六个文档读完了之后我们再对一下；

**Comment by ryang-max** [comment]
update:
写了半天serve，感觉好像又不是很需要，可能sglang native server api就够用了吧。我直接在jin的codebase上边跑边调了，先看看遗留问题多不多。。。

**Comment by zhaochenyang20** [comment]
> update: 写了半天serve，感觉好像又不是很需要，可能sglang native server api就够用了吧。我直接在jin的codebase上边跑边调了，先看看遗留问题多不多。。。

你的理解是正确的，就用 native server，不用 verl server

**Comment by zhaochenyang20** [comment]
> > update: 写了半天serve，感觉好像又不是很需要，可能sglang native server api就够用了吧。我直接在jin的codebase上边跑边调了，先看看遗留问题多不多。。。
> 
> 你的理解是正确的，就用 native server，不用 verl server

**Comment by ryang-max** [comment]
update: 
Successfully ran trl train with sglang backend in 2 gpus(1 for sglang server, 1 for trl-train)
Temporary TODO:
- [ ] Verify the correctness of training result
- [ ] Check if current implementation is as expected
- [ ] Test on multi-gpu/multi-node 

cc @renxinx @zhaochenyang20

**Comment by zhaochenyang20** [comment]
> 4.22 update: Successfully ran trl train with sglang backend in 2 gpus(1 for sglang server, 1 for trl-train): branch: https://github.com/ryang-max/trl/tree/sglang-server steps:
> 
> ```
> # start sglang-server
> python3 -m sglang.launch_server --model-path qwen/qwen2.5-0.5b-instruct
> 
> # modify cuda config in run.sh and grpo_config.yaml
> # run script
> sh ./trl/scripts/grop_test/run.sh
> ```
> 
> Temporary TODO (ETA Thu):
> 
> * [ ]  Verify the correctness of training result[ ]  Check if current implementation is as expected[ ]  Test on multi-gpu/multi-node
> 
> cc [@renxinx](https://github.com/renxinx) [@zhaochenyang20](https://github.com/zhaochenyang20)

模型太小了，建议用 7B 模型，反正我们出的了这么多卡。不用测试 multi-node，没人会用 trl 搞 multi-node，其他的真棒！

**Comment by jhinpan** [comment]
理论上是要跟现在的vLLM对齐精度？就跟我们之前VeRL一样，但是看大家有没有时间，我下周可以来一起帮忙比如说我们就现在gsm8k上对齐一个GRPO的精度看看有没有什么问题做一些benchmark。

**Comment by zhaochenyang20** [comment]
是的，这个精度你之前对过么？

**Comment by jhinpan** [comment]
没有印象了，上次对精度应该还是跟NV他们对Dpsk的精度的时候，在IF-Eval上。

**Comment by ryang-max** [comment]
4.23 update：
https://github.com/ryang-max/trl/tree/sglang-server
这里基本把from_disk的跑通了，能够和vllm跑同样任务loss对的上。还有一些小问题：
1. 算logits时候报了莫名的错挂掉了，感觉跟我们的代码应该没啥关系？我明天尝试复现下

![Image](https://github.com/user-attachments/assets/301deeaf-8ba1-4a1a-87d4-9c8df2dcf477)
2. 速度太慢了。和vllm用同样的配置（1卡inference，3卡train），预计速度大约慢3倍。我觉得可能是我对多卡训练的理解有问题，model save/load太多次了（https://github.com/ryang-max/trl/blob/8debe2abe7e4dbebf121c800caf8d9a73df770e7/trl/trainer/grpo_trainer.py#L931-L941）。对于每个进程，没开deepspeed的情况下我是需要每个都save一遍吗？还是应该怎么聚合一下在main_process里save一次？这块我自己有些没理解，另外对vllm的update实现其实我也没看太明白，但我感觉他们一开始的实现好像就是udpate_from_tensor的，毕竟没有save model。这块求助下 @jhinpan 
3. 我启训练的时候脚本里声明了wandb.init()，但是在vllm那里就只生成了一个run，sglang这里我用几张卡就生成了几个run（虽然只有一个里面有数据），脚本都是一样的，这个没太理解

cc @renxinx

**Comment by zhaochenyang20** [comment]
1. 我现在看这个 debug 很难搞，建议你写 dev-log。如果好几天你都解决不了一个 bug，就停下来好好想一想，捋一下自己那儿写的有问题。
2. 这个速度慢是非常不合理的，但是你想的原因是合理的，检查下 vllm 的参数更新逻辑，重新捋明白了再开始，不会缺任何接口的。
3. 这个可以扔到 verl 群里问，我回头拉你

**Comment by zhaochenyang20** [comment]
我们要在 trl 的 GRPO trainer 里面支持 SGLang，目前把这些文档顺着读完就好了。

1. https://zhuanlan.zhihu.com/p/5220718268
2. https://zhuanlan.zhihu.com/p/677607581
3. https://github.com/zhaochenyang20/Awesome-ML-SYS-Tutorial/tree/main/rlhf/OpenRLHF#rlhf-%E7%9A%84%E8%AE%A1%E7%AE%97%E6%B5%81
4. https://zhuanlan.zhihu.com/p/16370000391
5. https://github.com/zhaochenyang20/Awesome-ML-SYS-Tutorial/blob/main/rlhf/OpenRLHF/develop-log.md
6. https://github.com/zhaochenyang20/Awesome-ML-SYS-Tutorial/blob/main/rlhf/GRPO/SGLang_GRPO.md
7. https://github.com/zhaochenyang20/Awesome-ML-SYS-Tutorial/blob/main/rlhf/verl/server-based/veRL-server-based-rollout.md

TODO:
- [x] run trl[vllm] @ryang-max @renxinx 这个都得跑，而且瑞国需要交接好怎么跑 ETA: 4.18
- [ ] implement sglang-serve(server-based sglang for trl, similar to vllm) @ryang-max 这个并不麻烦，非常好写 ETA 4.21
- [ ] implement grpo_config and grpo_train @renxinx @ryang-max ETA 4.23
- [ ] Integration Test @renxinx ETA 4.26
- [ ] Compare performance with vllm @ryang-max @renxinx ETA 4.29

**Comment by zhaochenyang20** [comment]
1. 完善 HuggingFace 的人提到的一切 reivew https://github.com/huggingface/trl/pull/3370
2. 把在 trl 上使用 sglang 做 grpo 的文档附上去，要 merge 给 trl
3. 修复 dev-log，现在写的大量内容过时了

https://github.com/zhaochenyang20/Awesome-ML-SYS-Tutorial/blob/main/rlhf/GRPO/SGLang_GRPO.md

**Comment by renxinx** [comment]
1. profiling/Time taken: GRPOTrainer.compute_loss &  train/loss 上述两个的物理意义，为什么一个在低一个在高，是否符合预期 ⭐️ 
2. train/kl 上涨是否符合预期 ⭐️ 
3. dl中train test validation三个subset的意义 ⭐️ 
4. 写类似文档，教会大家如何配环境以及跑实验 https://github.com/volcengine/verl/blob/zhaochenyang20-patch-1/recipe/sppo/README.md#reproduce-the-experiment 💫 



以上3点搞明白后，对一下：
1. 怎么在trl中使用validation 💫 
3. train/grad_norm下降是否符合预期 
4. train/completions/clipped_ratio物理意义，什么是好的

**Comment by zhaochenyang20** [comment]
谁来把这个修了！

https://github.com/huggingface/trl/pull/3370

**Comment by Qihang-Zhang** [comment]
我来试一下🙋🏻‍♀️

**Comment by zhudianGG** [comment]
我也来试试，欢迎交流

**Comment by Luoxiaogan** [comment]
So do I.🤠

**Comment by PrinsYin** [comment]
我也来试试 🙌🙌

**Comment by zhaochenyang20** [comment]
> Hi Chenyang,
> 
> 读了下link里贴出的verl相关的文档，对Part1里提到的SPMD模式有点疑问，文档中提到"mock的SPMD已经足够满足性能需求"，但似乎没有详细说明: 真正的SPMD和mock SPMD的性能差异以及在什么情况下需要考虑真正的SPMD，另外或许文档里少了些框架之间性能对比以及内存占用的情况的测试数据 Part2里我这边看起来，FSDP的offloading在数据传输开销和分片上或许还有些优化空间在的，文档里要是有些结合相应场景的实践就更好了
> 
> 也欢迎其他朋友一起讨论和分享哈
> 
> [@zhaochenyang20](https://github.com/zhaochenyang20)

很多都东西确实得做实验验证，这个是我们讨论的一些结果。单独抽出来验证还挺难的

欢迎贡献？
